### DISCUSSION TITLE
Understanding CLAASP methods

### DISCUSSION THREAD

**Discussion by Cybro0** [Crypto-TII/claasp#139] [Q&A]
Hello, so im working on Linear and Differential trails. Any help like how to use these functions mentioned in CLAASP and how data should be given in it to perform linear and differential cryptanalysis.?

**Comment by juaninf** [comment]
Hi @Cybro0, Here's an example demonstrating how to use CLAASP for finding differential trails:

`https://github.com/peacker/claasp_white_paper/blob/main/section_4.3/speck_128_128_unsat_48.sage`

For examples on employing CLAASP to discover linear trails, check out:

`https://github.com/peacker/claasp_white_paper/blob/main/section_4.3/new_salsa_linear_trail.sage`

Additionally, the library contains numerous examples for differential search using SAT techniques, located at:

`tests/unit/cipher_modules/models/sat/sat_models/sat_xor_differential_model_test.py`

For searching linear trails with SAT techniques, visit:

`tests/unit/cipher_modules/models/sat/sat_models/sat_xor_linear_model_test.py`

Should you be interested in exploring other techniques like MILP, CP, or SMT, you'll find ample examples under `tests/unit/cipher_modules/models/`. For further documentation, please refer to `https://claasp.readthedocs.io/en/latest/`

**Reply by Cybro0** [reply]
Thank you for the examples. What does the `total_weight` represents at the resultant output? and can we directly run these functions by the cipher description.? Just need to know what parameters have to be changed.

**Reply by juaninf** [reply]
What does the `total_weight` represents at the resultant output? 
It is a variable related to the concept of probability/correlation weight in linear or differential trails. Thus, `total_weight` is the weight of an entire trail.

**Reply by Cybro0** [reply]
ok, assuming its giving me a trail by its cipher description. How to know which bits are involved in the trail and how to further calculate bias value from it?

**Reply by juaninf** [reply]
For example. This is the description of speck32/64 reduced to 2 rounds

```
{'cipher_id': 'speck_p32_k64_o32_r2', 'cipher_type': 'block_cipher', 'cipher_inputs': ['plaintext', 'key'], 'cipher_inputs_bit_size': [32, 64], 'cipher_output_bit_size': 32, 'cipher_number_of_rounds': 2, 'cipher_rounds': [[{'id': 'rot_0_0', 'type': 'word_operation', 'input_bit_size': 16, 'input_id_link': ['plaintext'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]], 'output_bit_size': 16, 'description': ['ROTATE', 7]}, {'id': 'modadd_0_1', 'type': 'word_operation', 'input_bit_size': 32, 'input_id_link': ['rot_0_0', 'plaintext'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], [16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]], 'output_bit_size': 16, 'description': ['MODADD', 2, None]}, {'id': 'xor_0_2', 'type': 'word_operation', 'input_bit_size': 32, 'input_id_link': ['modadd_0_1', 'key'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], [48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63]], 'output_bit_size': 16, 'description': ['XOR', 2]}, {'id': 'rot_0_3', 'type': 'word_operation', 'input_bit_size': 16, 'input_id_link': ['plaintext'], 'input_bit_positions': [[16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]], 'output_bit_size': 16, 'description': ['ROTATE', -2]}, {'id': 'xor_0_4', 'type': 'word_operation', 'input_bit_size': 32, 'input_id_link': ['xor_0_2', 'rot_0_3'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]], 'output_bit_size': 16, 'description': ['XOR', 2]}, {'id': 'intermediate_output_0_5', 'type': 'intermediate_output', 'input_bit_size': 16, 'input_id_link': ['key'], 'input_bit_positions': [[48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63]], 'output_bit_size': 16, 'description': ['round_key_output']}, {'id': 'intermediate_output_0_6', 'type': 'intermediate_output', 'input_bit_size': 32, 'input_id_link': ['xor_0_2', 'xor_0_4'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]], 'output_bit_size': 32, 'description': ['round_output']}], [{'id': 'constant_1_0', 'type': 'constant', 'input_bit_size': 0, 'input_id_link': [''], 'input_bit_positions': [[]], 'output_bit_size': 16, 'description': ['0x0000']}, {'id': 'rot_1_1', 'type': 'word_operation', 'input_bit_size': 16, 'input_id_link': ['key'], 'input_bit_positions': [[32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47]], 'output_bit_size': 16, 'description': ['ROTATE', 7]}, {'id': 'modadd_1_2', 'type': 'word_operation', 'input_bit_size': 32, 'input_id_link': ['rot_1_1', 'key'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], [48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63]], 'output_bit_size': 16, 'description': ['MODADD', 2, None]}, {'id': 'xor_1_3', 'type': 'word_operation', 'input_bit_size': 32, 'input_id_link': ['modadd_1_2', 'constant_1_0'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]], 'output_bit_size': 16, 'description': ['XOR', 2]}, {'id': 'rot_1_4', 'type': 'word_operation', 'input_bit_size': 16, 'input_id_link': ['key'], 'input_bit_positions': [[48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63]], 'output_bit_size': 16, 'description': ['ROTATE', -2]}, {'id': 'xor_1_5', 'type': 'word_operation', 'input_bit_size': 32, 'input_id_link': ['xor_1_3', 'rot_1_4'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]], 'output_bit_size': 16, 'description': ['XOR', 2]}, {'id': 'rot_1_6', 'type': 'word_operation', 'input_bit_size': 16, 'input_id_link': ['xor_0_2'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]], 'output_bit_size': 16, 'description': ['ROTATE', 7]}, {'id': 'modadd_1_7', 'type': 'word_operation', 'input_bit_size': 32, 'input_id_link': ['rot_1_6', 'xor_0_4'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]], 'output_bit_size': 16, 'description': ['MODADD', 2, None]}, {'id': 'xor_1_8', 'type': 'word_operation', 'input_bit_size': 32, 'input_id_link': ['modadd_1_7', 'xor_1_5'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]], 'output_bit_size': 16, 'description': ['XOR', 2]}, {'id': 'rot_1_9', 'type': 'word_operation', 'input_bit_size': 16, 'input_id_link': ['xor_0_4'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]], 'output_bit_size': 16, 'description': ['ROTATE', -2]}, {'id': 'xor_1_10', 'type': 'word_operation', 'input_bit_size': 32, 'input_id_link': ['xor_1_8', 'rot_1_9'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]], 'output_bit_size': 16, 'description': ['XOR', 2]}, {'id': 'intermediate_output_1_11', 'type': 'intermediate_output', 'input_bit_size': 16, 'input_id_link': ['xor_1_5'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]], 'output_bit_size': 16, 'description': ['round_key_output']}, {'id': 'cipher_output_1_12', 'type': 'cipher_output', 'input_bit_size': 32, 'input_id_link': ['xor_1_8', 'xor_1_10'], 'input_bit_positions': [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]], 'output_bit_size': 32, 'description': ['cipher_output']}]], 'cipher_reference_code': None}
```

And this is the output of a 2-round differential trail for speck32/64

```
{'cipher_id': 'speck_p32_k64_o32_r2', 'model_type': 'xor_differential', 'solver_name': 'kissat', 'solving_time_seconds': 0.0, 'memory_megabytes': 275.43552, 'components_values': {'plaintext': {'value': '00102000'}, 'key': {'value': '0000000000000000'}, 'rot_0_0': {'value': '2000', 'weight': 0}, 'modadd_0_1': {'value': '0000', 'weight': 1}, 'xor_0_2': {'value': '0000', 'weight': 0}, 'rot_0_3': {'value': '8000', 'weight': 0}, 'xor_0_4': {'value': '8000', 'weight': 0}, 'intermediate_output_0_5': {'value': '0000', 'weight': 0}, 'intermediate_output_0_6': {'value': '00008000', 'weight': 0}, 'constant_1_0': {'value': '0000', 'weight': 0}, 'rot_1_1': {'value': '0000', 'weight': 0}, 'modadd_1_2': {'value': '0000', 'weight': 0}, 'xor_1_3': {'value': '0000', 'weight': 0}, 'rot_1_4': {'value': '0000', 'weight': 0}, 'xor_1_5': {'value': '0000', 'weight': 0}, 'rot_1_6': {'value': '0000', 'weight': 0}, 'modadd_1_7': {'value': '8000', 'weight': 0}, 'xor_1_8': {'value': '8000', 'weight': 0}, 'rot_1_9': {'value': '0002', 'weight': 0}, 'xor_1_10': {'value': '8002', 'weight': 0}, 'intermediate_output_1_11': {'value': '0000', 'weight': 0}, 'cipher_output_1_12': {'value': '80008002', 'weight': 0}}, 'total_weight': 1.0, 'status': 'SATISFIABLE', 'building_time_seconds': 0.0016567707061767578, 'test_name': 'find_lowest_weight_xor_differential_trail'}
```

_How to know which bits are involved in the trail_

The bits with active differences in the component `plaintext` are those with value 1 in the binary representation of '0x00102000'. Similarly, the bits with active differences in the output of the trail (i.e., the `component cipher_output_1_12`)  are those with value 1 in the binary representation of '0x80008002'. You can also use the class Report to obtain the differences with a certain format. Here is an example of how to use the Report class

```
from claasp.cipher_modules.models.utils import get_single_key_scenario_format_for_fixed_values
from claasp.cipher_modules.report import Report
from claasp.ciphers.block_ciphers.speck_block_cipher import SpeckBlockCipher
from claasp.cipher_modules.models.sat.sat_models.sat_xor_differential_model import SatXorDifferentialModel
speck = SpeckBlockCipher(block_bit_size=32, key_bit_size=64, number_of_rounds=2)
print(speck.as_python_dictionary())
sat = SatXorDifferentialModel(speck)
trail = sat.find_lowest_weight_xor_differential_trail(get_single_key_scenario_format_for_fixed_values(speck), solver_name="kissat")
print(trail)
repo = Report(speck, trail)
repo.print_report()
```

The output is

```
        plaintext       
_ _ _ _ _ _ _ _ _ _ _ 1 _ _ _ _ _ _ 1 _ _ _ _ _ _ _ _ _ _ _ _ _ 

        intermediate_output_0_6 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 1 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

        cipher_output_1_12      
1 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 1 _ _ _ _ _ _ _ _ _ _ _ _ _ 1 _ 


KEY FLOW

        key     
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

        intermediate_output_0_5 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

        intermediate_output_1_11        
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

```

Here, _ means 0, while 1 is an active difference. Also, `intermediate_output_0_6` is the output of the first round. `key` is with all 0s (_) since the search was in the single-key scenario. Notice that for `key` we have 64 (_) since the key in speck32/64 is 64, and all of them are 0 because we search the trail in the single-key scenario.

_how to further calculate bias value from it?_
The bias is already estimated for CLAASP through the probability. In this example, the probability is 2^-total_weight = 2^-1

**Reply by Cybro0** [reply]
ok, thank you for the information. But, what is single key scenario and are there any other scenarios other than single-key scenario?

**Reply by juaninf** [reply]
Yes, other one for example is the related-key scenario. Which cipher are you analyzing?

**Reply by Cybro0** [reply]
Im trying on Speck itself, to understand how i can use CLAASP for linear and differential cryptanalysis. What is the difference by using single key and related key scenarios? and by just giving above mentioned lines i can find total_weight and the bits in trail. is there any other methods supporting for both cryptanalysis like data complexity or just have to calculate through this.?

**Reply by juaninf** [reply]
In the single-key scenario, the attacker cannot leverage previous keys to extract information about the cipher, unlike in the related-key scenario. Therefore, in the single-key scenario, the difference constraint for the key is set to zero. Conversely, in the related-key scenario, the attacker is permitted to utilize various older keys. This enables the attacker to attempt to establish the relationship between these keys by observing a pattern between these older keys (for example, the difference among them). [Here](https://ioactive.com/wp-content/uploads/2015/07/ldc_tutorial.pdf) a tutorial to mount an attack on a sbox toy cipher in the single-key scenario. You can implement that cipher on CLAASP and reproduce the results of that tutorial.

**Reply by Cybro0** [reply]
Is there any other methods supporting for both cryptanalysis like to know data complexity or just have to calculate through this.?

**Reply by juaninf** [reply]
The approach you take depends on the type of attack you are considering. If you aim to determine whether a cipher can be distinguished from a Pseudo Random Permutation (PRP), the basic guideline is that the data complexity for a differential attack is 2 to the power of negative total weight, and for a linear attack, it is 2 to the power of negative two times the total weight. On the other hand, if you intend to carry out a key-recovery attack, you will need to follow additional steps that are not yet supported in claasp but are intended to be included in its development roadmap.

**Reply by Cybro0** [reply]
If we give any model such as SAT or MILP, can we get to see by which approach they are following for the diff/linear cryptanalysis as we're just passing it by its cipher description.? and also does different models gives different values? such as the total_weight or trail, varies according to the type of model?

**Reply by juaninf** [reply]
> If we give any model such as SAT or MILP, can we get to see by which approach they are following for the diff/linear cryptanalysis as we're just passing it by its cipher description.?
Sorry, I did not get your question. could you please give me an example?
> and also does different models give different values?
it depends, if you are searching for optimal trails, they should give you the same `total_weight`. But not necessarily the same trail, since could be many trails with the same `total_weight`.

**Comment by Cybro0** [comment]
I'm getting this error for the generate csv report. Same syntax u given in the claasp documentation.

![image](https://github.com/Crypto-TII/claasp/assets/143835501/8600e4e6-a96e-47a9-8308-a162bf1d539d)

**Reply by juaninf** [reply]
Hi @Cybro0, the cipher class (in this case Identity) does not have the method `generate_csv_report`. Could you please send the link of the documentation where you found the error? What do you want to do? Can I help you?

**Reply by Cybro0** [reply]
I was just exploring other methods that are present in CLAASP. However, keeping that aside. coming back to differential and linear cryptanalysis. Im running differential one as u mentioned example above. Its working. But for the same , Im running linear models its taking so much time that still its running and not getting any output. Is there any such example for Linear trail like u provided for the differential one.? like which gives me quicker results.

**Reply by juaninf** [reply]
could you please provide a minimal example code you are using to search for linear trails?
